### PR TITLE
Happy Card Writeup

### PR DESCRIPTION
**Pull request by rad4day**
@I-Al-Istannen pls look

### CODE REVIEW

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+0x80 0xB8 0x00 0x00 0x08 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x00 0x7F;
+0x00 0xA4 0x04 0x00 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x7F;
+0x88 0x88 0x00 0x00 0x30 $FLAG 0x7f;
+0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;
+EOF
+
+TMPDIR=/jctmp
+mkdir $TMPDIR
+cd /upload
+for capfile in *.cap; do
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:82]:
```suggestion
for capfile in *.cap; do # 7
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+	java -Djc.home="$JC_HOME_TOOLS" -classpath "$TOOLS_CP" com.sun.javacard.offcardverifier.Verifier -nobanner $@
+}
+
+scriptgen() {
+	java -Djc.home="$JC_HOME_SIM" -classpath "$SIM_CP" com.sun.javacard.scriptgen.Main -nobanner $@
+}
+
+script=/tmp/script
+
+verifycap -outfile /tmp/hello.cap.digest /files/hello.cap
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:63]:
```suggestion
verifycap -outfile /tmp/hello.cap.digest /files/hello.cap # 3
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+scriptgen -hashfile /tmp/hello.cap.digest -o /tmp/hello.cap.script /files/hello.cap
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:64]:
```suggestion
scriptgen -hashfile /tmp/hello.cap.digest -o /tmp/hello.cap.script /files/hello.cap # 3
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+cat << EOF > $script
+powerup;
+output off;
+0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;
+EOF
+FLAG=`echo -n "$FLAG"|perl -lne 'print map {"0x".(unpack "H*",$_)." "} split //, $_;'`
+cat /tmp/hello.cap.script >> $script
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:71]:
```suggestion
cat /tmp/hello.cap.script >> $script # 3
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+cat << EOF >> $script
+0x80 0xB8 0x00 0x00 0x08 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x00 0x7F;
+0x00 0xA4 0x04 0x00 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x7F;
+0x88 0x88 0x00 0x00 0x30 $FLAG 0x7f;
+0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:76]:
```suggestion
0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F; # 6
```
Is this the correct location?

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+cat << EOF > $script
+powerup;
+output off;
+0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;
+EOF
+FLAG=`echo -n "$FLAG"|perl -lne 'print map {"0x".(unpack "H*",$_)." "} split //, $_;'`
+cat /tmp/hello.cap.script >> $script
+cat << EOF >> $script
+0x80 0xB8 0x00 0x00 0x08 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x00 0x7F;
+0x00 0xA4 0x04 0x00 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x7F;
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:74]:
```suggestion
0x00 0xA4 0x04 0x00 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x7F; # 4
```
Is this the correct location?

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+}
+
+script=/tmp/script
+
+verifycap -outfile /tmp/hello.cap.digest /files/hello.cap
+scriptgen -hashfile /tmp/hello.cap.digest -o /tmp/hello.cap.script /files/hello.cap
+cat << EOF > $script
+powerup;
+output off;
+0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:68]:
```suggestion
0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F; # 2
```
Is this the correct location?

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+---
+
+[Real World CTF 2023](https://realworldctf.com/challenge) was a jeopardy-style capture-the-flag event organized by Chaitin Tech.
+We participated as part of the [Sauercloud](https://ctftime.org/team/54748) CTF-team.
+
+## How solve? Where exploit?
+
+- Google for java card stuff
+- Find [some usable tooling to build your own applets](https://github.com/martinpaljak/ant-javacard)
+- Enable eeprom dumping in cref (`-o /upload/eeprom`)
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:15]:
```suggestion
- Enable EEPROM dumping in cref (`-o /upload/eeprom`)
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+- Stumble upon [Sergei Volokitin's Master's Thesis](https://www.ru.nl/publish/pages/769526/sergei_volokitin.pdf) for an overview of vulnerabilities
+    - Try multiple, but be caught by byte code verification / type checking
+- Stumble upon [PhiAttack](https://cardis2021.its.uni-luebeck.de/papers/CARDIS2021_Dubreuil.pdf)
+    - Needs older sdk
+    - Implement simple type confusion between `Object` and `byte[]`
+- Be confused why some casts work and others dont
+    - reverse cref `__saload` and `__baload` read checks
+- Use type confusion to cast an `Object` to a `byte[]` array with arbitrary length
+- Read eeprom data using (now no longer) out-of-bounds read on `byte[]`
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:24]:
```suggestion
- Read EEPROM data using (now no longer) out-of-bounds read on `byte[]`
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+sleep 5
+java -Djc.home="$JC_HOME_SIM" -classpath "$SIM_CP" com.sun.javacard.apdutool.Main -nobanner -noatr $script
+```
+
+The APDU commands in this script do the following:
+1. disable output
+2. switch to installer applet
+3. upload hello.cap
+4. switch to hello applet
+5. insert flag into eeprom
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:107]:
```suggestion
5. insert flag into EEPROM
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+        }
+    }
+}
+
+```
+
+## EEPROM Object Layout
+![](/imgs/rwctf23-happycard-imhex.png)
+
+We've taken the liberty to create a small imHex-Pattern to highlight a few of the important bits of an eeprom dump we obtained during analysing of the challenge.
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:182]:
```suggestion
We've taken the liberty to create a small imHex-Pattern to highlight a few of the important bits of an EEPROM dump we obtained during analysing of the challenge.
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+
+The first highlighted block is a `byte[]` filled with `A`s,  directly followed by the definition of our ExploitApplet.
+
+The second highlighted block is a custom Object we named `Foo`, containing a field `short len = 0x1337`.
+
+And the final block is the FlagArray we want to read, which isn't completely straight forward, as the java card simulator implements a few features to isolate our ExploitApplet from the Safe.
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:188]:
```suggestion
And the final block is the flag array we want to read, which isn't completely straight forward, as the java card simulator implements a few features to isolate our ExploitApplet from the Safe.
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+
+### Object references
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:190]:
```suggestion
### Object References
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+  inflating: attachments/files/entrypoint.sh
+  inflating: attachments/files/hello.cap
+  inflating: attachments/files/java_card_simulator-3_1_0-u5-win-bin-do-b_70-09_mar_2021.msi
+  inflating: attachments/files/java_card_tools-win-bin-b_17-06_jul_2021.zip
+  inflating: attachments/start.sh
+```
+
+So a pretty basic docker setup as well as a `hello.cap`, and a few tools related to `java-card` including the simulator in version 3.1.0u5.
+
+## The entrypoint
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:49]:
```suggestion
## The Entrypoint
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+layout: post
+title: "Real World CTF 2023: Happy-Card Writeup"
+categories: writeups
+author: I Al Istannen, rad4day
+---
+
+[Real World CTF 2023](https://realworldctf.com/challenge) was a jeopardy-style capture-the-flag event.
+We participated as part of the [Sauercloud](https://ctftime.org/team/54748) CTF-team.
+
+## How solve? Where exploit?
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:11]:
```suggestion
## How Solve? Where Exploit?
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+    u8 securityContext = 0x22;
+    u16 arrayType = 0x0;       // this is a byte array?
+    u8 arrayMagic = 0x1C;      // this is an array
+    u16 length = 0x30;
+    // end of header, followed by data
+    u8 data[length];
+};
+```
+
+Note that the FlagArray has a different `securityContext`.
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:212]:
```suggestion
Note that the flag array has a different `securityContext`.
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+### The PhiAttack
+
+To circumvent the verification, we simultaneously need to produce a congruent export file and also import the methods as they were exported.
+This is not possible with two cap files, but we aren't constrained to just two.
+
+By creating a layout similar to the following (from the paper linked above):
+
+![](/imgs/rwctf23-happycard-phiattack.png)
+
+we can trick the verifier into accepting all cap files, while *still* having a method performing a type confusion.
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:283]:
```suggestion
we can trick the verifier into accepting all CAP files, while *still* having a method performing a type confusion.
```

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+- `aaload` opcode checks that you read from a reference array
+- `saload` opcode checks that you read from a short array
+- `iaload` opcode checks that you read from an int array
+- `baload` opcode checks that you are not reading from any of the above. <strong style="font-weight: bold;">wat</strong>.
+
+Therefore, we are free to read bytes from *anything*, provided it is not a `short`, `int` or `Object` array.
+
+Very conveniently, the `Object` header is 2 byte shorter than the `array` header. This allows us to freely control the `length` field, if we manage to interpret an `Object` as an array :)
+
+## The code
```
Comment (intrigus-lgtm) [inline _posts/2023-01-08-rwctf2023-happycard.md:298]:
```suggestion
## The Code
```

**Review by sevenmaster — APPROVED** [review]
accept writing improvments

File: _posts/2023-01-08-rwctf2023-happycard.md [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+        }
+    }
+}
+
+```
+
+## EEPROM Object Layout
+![](/imgs/rwctf23-happycard-imhex.png)
+
+We've taken the liberty to create a small imHex-Pattern to highlight a few of the important bits of an EEPROM dump we obtained during analysing of the challenge.
```
Comment (rad4day) [inline _posts/2023-01-08-rwctf2023-happycard.md:182]:
```suggestion
We've highlighted a few of the important bits of an EEPROM dump we obtained during analysing of the challenge using the ImHex Hex-Editor.
```

**Pull request merged**